QRServer/common/clienthandler.py

```diff
@@ -50,30 +48,50 @@ def _socket_reader(self):
                     self.in_queue.put(i)
 
     def register_handler(self, prefix: bytes, handler):
+        """Deprecated, do not use"""
         if prefix in self.handlers:
             self.handlers[prefix].append(handler)
         else:
             self.handlers[prefix] = [handler]
 
+    def register_message_handler(self, mtype: Type[RMT], handler: Callable[[RMT], None]):
+        if mtype in self.message_handlers:
+            self.message_handlers[mtype].append(handler)
+        else:
+            self.message_handlers[mtype] = [handler]
+
     def run(self):
         while self.cs:
             data = self.in_queue.get(block=True)
             values = data.split(messages.delim.encode('ascii'))
-            log.debug('handling: {}'.format(values))
             prefix = values[0]
 
-            if prefix in self.handlers:
-                try:
+            message = RequestMessage.from_data(data)
+            try:
+                if message is not None:
+                    log.debug('Handling: {}'.format(message))
+                    mtype = type(message)
+                    if mtype in self.message_handlers:
+                        for handler in self.message_handlers[mtype]:
+                            handler(message)
+                    else:
+                        log.error('No handler for message type {}'.format(mtype))
+                elif prefix in self.handlers:
```

Comment:
is there any realistic way for this to be executed?

Reply:
In the game, we still handle `<S>` the old way because I'm not sure if I got every possible packet described :)